### PR TITLE
SAK-49998 Assignments improve performance and UX in assignments by student view 

### PR DESCRIPTION
**Pull request by stetsche**
https://sakaiproject.atlassian.net/browse/SAK-49998

### CODE REVIEW

File: assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm [hunk 1]
```diff
@@ -28,9 +59,6 @@
 				#end
 				#viewFilterPanel("viewFilterForm", "viewgroup", $viewSelectOptions, "doChange_submission_list_option")
 			#end
-			#if( !$hasAtLeastOneAnonAssignment )
-				#searchFilterPanel("$form_search", $searchString, "doView_submission_list_search", "doView_submission_list_search_clear")
-			#end
```
[the remark this answers is not in the document]
Comment (stetsche) [reply assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm:33]: This removes the search field, which is replaced by the datatables search. I'll check if we can remove the context variable, or if it's used elsewere

File: velocity/tool/src/templates/VM_chef_library.vm [hunk 2]
```diff
@@ -705,3 +705,51 @@ ${extra}")
 #set($libLink="#libraryLink('js/sakai-table-toolbar/searchFilterPanelMacro.js')")
 <script type='text/javascript' src='$libLink' defer></script>
 #end ## searchFilterPanel
+
+#* -----------------------------------------------------------------------------------
+#  Adds the an object with a common datatables config
```
Comment (stetsche) [inline velocity/tool/src/templates/VM_chef_library.vm:710]: Fix this comment's grammar.

File: velocity/tool/src/templates/VM_chef_library.vm [hunk 3]
```diff
@@ -705,3 +705,51 @@ ${extra}")
...
+    },
+    placeholder: {
+        first: "$resourceLoader.getString('datatables.custom.first')",
+        last: "$resourceLoader.getString('datatables.custom.last')",
+        next: "$resourceLoader.getString('datatables.custom.next')",
+        previous: "$resourceLoader.getString('datatables.custom.previous')",
+    },
+}
+#end
+<!-- end: assignment_macros.vm -->
```
[the remark this answers is not in the document]
Comment (stetsche) [reply velocity/tool/src/templates/VM_chef_library.vm:755]: Yes, I moved it here from assignment_macros.vm, I will remove the comment.

File: assignment/tool/src/webapp/js/assignmentsByStudent.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
+(() => {
+
+// Include datatables dependencies
+window.includeWebjarLibrary('datatables');
+window.includeWebjarLibrary('datatables-rowgroup');
+
+// Assignments By Students "global" namespace
+window.ABS = {
```
Comment (adrianfish) [inline assignment/tool/src/webapp/js/assignmentsByStudent.js:8]: I think you should change the name of this, tbh. Nobody will know what ABS means when they're inspecting in the console. It could be "Antilock Braking System" :). I've been trying to namespace stuff with sakai.assignments, or something similar.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 5]
```diff
@@ -16757,4 +16773,12 @@ private String getRateSubmissionTimeSpent(AssignmentSubmission submission) {
         }
         return assignmentService.intToTime(assigmentRateSpent);
     }
+    
+    private Optional<User> getUser(String userId) {
```
Comment (adrianfish) [inline assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: You could move this into the UserDirectoryService. Maybe use the method "getOptionalUser"? We could perhaps use that naming pattern as a consistent way of overloading existing methods like getUser.

**Review by adrianfish — CHANGES_REQUESTED** [review]

File: assignment/tool/src/webapp/js/assignmentsByStudent.js (hunk 4 above)
Comment (stetsche) [reply assignment/tool/src/webapp/js/assignmentsByStudent.js:8]: I thought "Adrian Benefits Sakai" was a quite clear acronym, but you are right, I'll namespace it a bit better.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 5 above)
Comment (stetsche) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: Yeah, I think it makes sense, but I didn't want to go so far. I'll do it.

Comment (bgarciaentornos) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: Maybe we could keep it inline in the stream, it's a tad uglier but doesn't need kernel updates

Comment (stetsche) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: I have a special "Master adjustments" commit on this branch, I would be fine with adding it there as well. Because I think this pattern makes sense and will be introduced at some point anyway. If we anticipate it now, we won't have duplicated code later. For 22 based branched `this::getOptionalUser` and master `userDirectoyService::getOptionalUser.` What do you think?

Comment (bgarciaentornos) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: It's fine. I would rephrase it as "Master master adjustments" though

Comment (ern) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:16777]: Agree with everyone here that adding a new method UserDirectoyService::getOptionalUser to return an Optional aka
`public Optional<User> getOptionalUser(String userId)`

I approve of the method name here, and the original method should be annotated as deprecated. (it's a shame that the return type is not part of a method signature)

this also has a positive side effect that the Exception UserNotDefinedException does not need to be thrown. TBH this was what returning NULL or Optional<> is for. An Exception is for communicating to the callee of the method with a return value other than Objects presence or absence is not to be trusted.

**Comment by ern** [comment]
Nice changes @stetsche

File: assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm (hunk 1 above)
[the remark this answers is not in the document]
Comment (stetsche) [reply assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm:33]: We might still need the variable and display a banner based on that condition.

File: assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 5 above)
Comment (stetsche) [reply assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: I did add the new method to UserDirectoryService, but I did not deprecate the original one yet

**Comment by stetsche** [comment]
I could address some of the comments, but I have to move to a different project now. One of my colleagues will finish this. We also have found some cases to fix when testing.

File: assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm [hunk 6]
```diff
@@ -259,5 +259,10 @@
 			<input type="hidden" name="sakai_csrf_token" value="$sakai_csrf_token" />
 		</form>
 	#end
+	#if ($expandedUserName)
+		<script>
+			window.location.href = window.location.href + "#" + "$expandedUserName";
```
[the remark this answers is not in the document]
Comment (JuanDavid102) [reply assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm:264]: Hi Stetsche, I did something similar in the code. If you think another change or some improvement tell me

[the remark this answers is not in the document]
Comment (stetsche) [reply assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm:264]: Sure, I will take a look again, when after you update the PR!

[the remark this answers is not in the document]
Comment (stetsche) [reply assignment/tool/src/webapp/vm/assignment/chef_assignments_instructor_student_list_submissions.vm:264]: Ahh, I see you created another PR!

**Review by bgarciaentornos — APPROVED** [review]
